### ISSUE TITLE
Unable to retrieve even simple Gazebo message via subscriber callback from develop branch?

### ISSUE THREAD

**Issue by Skylion007** [jpieper/pygazebo#27]
```
import trollius
from trollius import From

import pygazebo
import pygazebo.msg.world_stats_pb2
@trollius.coroutine
def publish_loop():
    manager = yield From(pygazebo.connect()) 
    print("hello")

    @trollius.coroutine
    def callback2(data):
        print(data)
        message = pygazebo.msg.world_stats_pb2.WorldStatistics.ParseFromString(data)
        print(message)

     subscriber = manager.subscribe('/gazebo/world/world_stats',
                     'gazebo.msgs.WorldStatistics',
                     callback2)

    subscriber.wait_for_connection()
    while(True):
        trollius.sleep(1)

loop = trollius.get_event_loop()
loop.run_until_complete(publish_loop())
```

I wanted to test some basic subscribers. I have been using publishers for two weeks or so and have gotten them working quite well. However, I cannot seem to get subscribers working at all. What am I doing wrong? Before you ask, my SDF has the default_world as world. I have checked and the topic '/gazebo/world/world_stats/' does actually have subscribers according to the gz tool. Unfortunately, no matter how hard I try, I can not get the python to run the callback2 at all. Even if I just have hello world, nothing happens. What am I doing wrong? Is there a bug in the current develop branch?

**Comment by Skylion007** [comment]
Never mind, my own idiocy. Forgot that Trollius.sleep() required a yield.

**Comment by RDaneelOlivav** [comment]
To me it gives the following when changing only the topic to "/gazebo/default/world_stats", which has the same msg type error:
No handlers could be found for logger "trollius"

And when added a logger at the top in imports:
import logging
logging.basicConfig()

Gives:

ERROR:trollius:Exception in callback <lambda>(<Future ...xbb\x01'>) at /usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py:198
handle: <Handle <lambda>(<Future ...xbb\x01'>) at /usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py:198>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/trollius-1.0.4-py2.7.egg/trollius/events.py", line 151, in _run
    self._callback(*self._args)
  File "/usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py", line 198, in <lambda>
    lambda future: self._handle_read(future, connection))
  File "/usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py", line 209, in _handle_read
    self._connect3(connection)
TypeError: _connect3() takes exactly 3 arguments (2 given)

Any ideas? Does it happen also to you?

**Comment by Skylion007** [comment]
I can't remember exactly what I did to fix that code but this code below does indeed work.

``` Python
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 21 15:40:57 2015

@author: skylion
"""

import trollius
from trollius import From

import pygazebo
import pygazebo.msg.world_stats_pb2
@trollius.coroutine
def publish_loop():
    manager = yield From(pygazebo.connect()) 

    def callback(data):
        message = pygazebo.msg.world_stats_pb2.WorldStatistics.FromString(data)        
        sim_time = message.sim_time
        import math
        sim_time = sim_time.sec + sim_time.nsec/math.pow(10, 9)
        print(sim_time)

    subscriber = manager.subscribe('/gazebo/world/world_stats',
                     'gazebo.msgs.WorldStatistics',
                     callback)

    yield From(subscriber.wait_for_connection())
    while(True):
        yield From(trollius.sleep(1.00))
        print('wait...')
    print(dir(manager))

import logging

logging.basicConfig()

loop = trollius.get_event_loop()
loop.run_until_complete(publish_loop())
```

**Comment by RDaneelOlivav** [comment]
Wooow, gona try it right now,  gimi a sec ;)

**Comment by RDaneelOlivav** [comment]
Gives the same error, and I have only changed the 
subscriber = manager.subscribe('/gazebo/world/world_stats',
                     'gazebo.msgs.WorldStatistics',
                     callback)
To 

subscriber = manager.subscribe('/gazebo/default/world_stats',
                     'gazebo.msgs.WorldStatistics',
                     callback)

3910.605
ERROR:trollius:Exception in callback <lambda>(<Future ...xee\x01'>) at /usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py:198
handle: <Handle <lambda>(<Future ...xee\x01'>) at /usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py:198>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/trollius-1.0.4-py2.7.egg/trollius/events.py", line 151, in _run
    self._callback(*self._args)
  File "/usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py", line 198, in <lambda>
    lambda future: self._handle_read(future, connection))
  File "/usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py", line 209, in _handle_read
    self._connect3(connection)
TypeError: _connect3() takes exactly 3 arguments (2 given)
wait...
wait...
wait....
wait...

and there it stays for eternity.

Any insight? Does it happen the same to you?

**Comment by RDaneelOlivav** [comment]
The error takes you to here "/usr/local/lib/python2.7/dist-packages/pygazebo-3.0.0_2014.1-py2.7.egg/pygazebo/pygazebo.py"

Where we see something that it dazzels me a bit:

def _connect3(self, future, connection):
    future.result()  # check for error

```
if not self._connection_future.done():
    self._connection_future.set_result(None)
future = connection.read_raw()
future.add_done_callback(
    lambda future: self._handle_read(future, connection))
```

def _handle_read(self, future, connection):
    data = future.result()
    if data is None:
        self._connections.remove(connection)
        if len(self._connections) == 0:
            self._connection_future = asyncio.Future()
        return

```
self.callback(data)
self._connect3(connection)
```

Shouldnt _connect3 have 2 inputs (connection and FUTURE) ?
Is there something in futures that I'm anaware of?

**Comment by jpieper** [comment]
Nope, that is a bug which was fixed in d9137078b87d41cea2b350c08a0c229f6b4dac72

What it does demonstrate is that I need to make a new release with updated gazebo bindings.  I will try to do so soonish.  In the meantime, you can get the "develop" branch from github and use it.
